File: 1050-construct-binary-search-tree-from-preorder-traversal/construct-binary-search-tree-from-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:

        if not preorder:
            return None
        
        root = TreeNode(preorder[0])
        stack = [root]

        for value in preorder[1:]:
            node = TreeNode(value)

            if value < stack[-1].val:
                stack[-1].left = node
            else:
                while stack and value > stack[-1].val:
                    last = stack.pop()
                
                last.right = node
            stack.append(node)
        
        return root

File: 1050-construct-binary-search-tree-from-preorder-traversal/construct-binary-search-tree-from-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode bstFromPreorder(int[] preorder) {

        if (preorder.length < 1) {
            return null;
        }

        TreeNode root = new TreeNode(preorder[0]);

        int mid = preorder.length;
        for (int i = 0; i < preorder.length; i++) {
            if (preorder[i] > preorder[0]) {
                mid = i;
                break;
            }
        }

        root.left = bstFromPreorder(Arrays.copyOfRange(preorder, 1, mid));
        root.right = bstFromPreorder(Arrays.copyOfRange(preorder, mid, preorder.length));

        return root;
        
    }
}